### PR TITLE
hadoop: 3.3.5 -> 3.3.6, build container executor from source

### PR DESCRIPTION
**Pull request by illustris**
## Description of changes

- hadoop: 3.3.5 -> 3.3.6
- [upgrade openssl dep from 1.x to 3.x](https://github.com/NixOS/nixpkgs/issues/210452)
    - build YARN container executor from source with openssl3
- simplify the needlessly complex directory structure of the hadoop package
    - directly extract tarball to $out instead of $out/lib/hadoop.x.x.x/
    - update module and test to account for the change
- switch to `finalAttrs` instead of rec, to simplify overriding attrs

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ConnorBaker** [comment]
@illustris any reason you requested my review? I don't have experience with Hadoop or Nixpkgs' handling of Hadoop :(

**Comment by illustris** [comment]
Hey, sorry if this PR is not relevant to you. I think GitHub recommends reviewers based on the edit history. You've previously contributed to aarch64 support for Hadoop and Spark in nixpkgs.

**Review by NukaDuka — APPROVED** [review]
Looks good to me.

**Comment by illustris** [comment]
That's weird... aarch64 build works for me, but failed on ofborg
@GrahamcOfBorg build hadoop

**Review by NukaDuka — APPROVED** [review]

File: pkgs/applications/networking/cluster/hadoop/containerExecutor.nix [hunk 1]
```diff
@@ -0,0 +1,37 @@
+{ version, stdenv, fetchurl, lib, cmake, openssl, platformAttrs, ... }:
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "hadoop-yarn-containerexecutor";
+  inherit version;
+
+  src = fetchurl {
+    url = with finalAttrs; "mirror://apache/hadoop/common/hadoop-${finalAttrs.version}/hadoop-${finalAttrs.version}-src.tar.gz";
```
Comment (pbsds) [inline pkgs/applications/networking/cluster/hadoop/containerExecutor.nix:8]:
```suggestion
    url = "mirror://apache/hadoop/common/hadoop-${finalAttrs.version}/hadoop-${finalAttrs.version}-src.tar.gz";
```

**Review by pbsds — COMMENTED** [review]
The darwin eval and build fixes strikes me as odd, but i am unable to test. Is there reason to believe this will work on darwin? Is containerExecutor an optional dependency?

**Comment by pbsds** [comment]
Result of `nixpkgs-review pr 256449` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>spark_3_2</li>
    <li>spark_3_3</li>
  </ul>
</details>
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>13 packages built:</summary>
  <ul>
    <li>hadoop (hadoop3 ,hadoop_3_3)</li>
    <li>hadoop2</li>
    <li>hadoop_3_2</li>
    <li>pig</li>
    <li>python310Packages.dask-yarn</li>
    <li>python310Packages.dask-yarn.dist</li>
    <li>python310Packages.skein</li>
    <li>python310Packages.skein.dist</li>
    <li>python311Packages.dask-yarn</li>
    <li>python311Packages.dask-yarn.dist</li>
    <li>python311Packages.skein</li>
    <li>python311Packages.skein.dist</li>
    <li>spark (spark3 ,spark_3_4)</li>
  </ul>
</details>

**Comment by illustris** [comment]
> The darwin eval and build fixes strikes me as odd, but i am unable to test. Is there reason to believe this will work on darwin? Is containerExecutor an optional dependency?

```
nix-repl> :lf .
warning: Git tree '/home/illustris/src/nixpkgs/.worktree/hadoop' is dirty
Added 16 variables.

nix-repl> legacyPackages.x86_64-darwin.hadoop
error:
       … while calling the 'derivationStrict' builtin

         at /builtin/derivation.nix:9:12: (source not available)

       … while evaluating derivation 'hadoop-3.3.6'
         whose name attribute is located at /nix/store/718xpf1lqi4pqfa6xlh738sklvjm984k-source/pkgs/stdenv/generic/make-derivation.nix:300:7

       … while evaluating attribute 'buildInputs' of derivation 'hadoop-3.3.6'

         at /nix/store/718xpf1lqi4pqfa6xlh738sklvjm984k-source/pkgs/stdenv/generic/make-derivation.nix:347:7:

          346|       depsHostHost                = lib.elemAt (lib.elemAt dependencies 1) 0;
          347|       buildInputs                 = lib.elemAt (lib.elemAt dependencies 1) 1;
             |       ^
          348|       depsTargetTarget            = lib.elemAt (lib.elemAt dependencies 2) 0;

       error: Package ‘libtirpc-1.3.3’ in /nix/store/718xpf1lqi4pqfa6xlh738sklvjm984k-source/pkgs/development/libraries/ti-rpc/default.nix:31 is not available on the requested hostPlatform:
         hostPlatform.config = "x86_64-apple-darwin"
         package.meta.platforms = [
           "aarch64-linux"
           "armv5tel-linux"
           "armv6l-linux"
           "armv7a-linux"
           "armv7l-linux"
           "i686-linux"
           "loongarch64-linux"
           "m68k-linux"
           "microblaze-linux"
           "microblazeel-linux"
           "mips-linux"
           "mips64-linux"
           "mips64el-linux"
           "mipsel-linux"
           "powerpc64-linux"
           "powerpc64le-linux"
           "riscv32-linux"
           "riscv64-linux"
           "s390-linux"
           "s390x-linux"
           "x86_64-linux"
         ]
         package.meta.badPlatforms = [ ]
       , refusing to evaluate.

       a) To temporarily allow packages that are unsupported for this system, you can use an environment variable
          for a single invocation of the nix tools.

            $ export NIXPKGS_ALLOW_UNSUPPORTED_SYSTEM=1

        Note: For `nix shell`, `nix build`, `nix develop` or any other Nix 2.4+
        (Flake) command, `--impure` must be passed in order to read this
        environment variable.

       b) For `nixos-rebuild` you can set
         { nixpkgs.config.allowUnsupportedSystem = true; }
       in configuration.nix to override this.

       c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
         { allowUnsupportedSystem = true; }
       to ~/.config/nixpkgs/config.nix.
```

libtirpc is not supported on Darwin. The commit that fixes the eval on darwin (https://github.com/NixOS/nixpkgs/pull/256449/commits/21fa56c145258730be931cb870094b87997429cf) just drops buildInputs from darwin because none of the native code in the upstream release is usable on mac anyway.

As for removing container executor for mac, it is a setuid linux binary used by the YARN Nodemanager to set up cgroups before starting a new container.  The hadoop (and spark) packages for mac in nixpkgs are meant to be used as clients to interact with clusters. You don't need container executor for that.

**Pull request merged**